src/librustc_mir/interpret/machine.rs

```diff
@@ -254,6 +254,14 @@ pub trait Machine<'mir, 'tcx>: Sized {
         kind: Option<MemoryKind<Self::MemoryKind>>,
     ) -> (Cow<'b, Allocation<Self::PointerTag, Self::AllocExtra>>, Self::PointerTag);
 
+    /// Called to notify the machine of deallocations
+    fn notify_deallocation(
+        _memory_extra: &Self::MemoryExtra,
+        _alloc: &Allocation<Self::PointerTag, Self::AllocExtra>,
+        _ptr: Pointer<Self::PointerTag>,
+    ) {
```

Comment:
While you are adding this, it is probably not a bad idea to allow it to fail, so please make the return type `InterpResult<'tcx>`.